File: Go-Solutions/nowcoder/nc27.go
package nowcoder

func subsets(num []int) [][]int {
	// write code here
	if len(num) == 0 {
		return [][]int{}
	}

	res := make([][]int, 0)
	list := make([]int, 0)
	dfs27(num, &res, &list, 0)

	return res
}

func dfs27(num []int, res *[][]int, list *[]int, begin int) {
	listToAdd := make([]int, len(*list))
	for i := 0; i < len(*list); i++ {
		listToAdd[i] = (*list)[i]
	}
	*res = append(*res, listToAdd)
	for i := begin; i < len(num); i++ {
		*list = append(*list, num[i])
		dfs27(num, res, list, i+1)
		*list = (*list)[:len(*list)-1]
	}
}

File: Java-Kotlin/src/main/java/nowcoder/nc27.java
package nowcoder;

import java.util.ArrayList;

public class nc27 {

    private static void dfs(int[] num, ArrayList<ArrayList<Integer>> res, ArrayList<Integer> list, int begin) {
        res.add(new ArrayList<>(list));
        for (int i = begin; i < num.length; i++) {
            list.add(num[i]);
            dfs(num, res, list, i + 1);
            list.remove(list.size() - 1);
        }
    }

    public ArrayList<ArrayList<Integer>> subsets(int[] num) {
        if (num == null || num.length == 0) {
            return new ArrayList<>();
        }

        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        ArrayList<Integer> list = new ArrayList<>();
        dfs(num, res, list, 0);

        return res;
    }
}
